lib/tokens.ts

```diff
@@ -0,0 +1,184 @@
+import { PublicKey } from "@solana/web3.js";
+import { TokenProps } from "./types";
+import {
+  JsonMetadata,
+  MPL_TOKEN_METADATA_PROGRAM_ID,
+  deserializeMetadata,
+} from "@metaplex-foundation/mpl-token-metadata";
+import { RpcAccount } from "@metaplex-foundation/umi";
+import { Provider } from "@coral-xyz/anchor";
+import { USDCAddress, USDCMetadata } from "./constants";
+import { Mint, getMint, getTokenMetadata } from "@solana/spl-token";
+/**
+ * Starts with the jup.ag strict list to find token. jup.ag maintains a list of quality tokens
+ * if that fails, use metaplex with RPC call and fetch metadata json
+ * if THAT fails, take MINT address and truncate by 5 characters and that's the symbol (no image)
+ */
+export async function enrichTokenMetadata(
+  tokenAddress: PublicKey,
+  rpcProvider: Provider
+): Promise<TokenProps> {
+  // first, is it USDC?
+  if (tokenAddress.toString() === USDCAddress) {
+    return USDCMetadata;
+  }
+
+  // second check jup list
+  const tokenOnJup = await getTokenFromJupStrictList(tokenAddress);
+  if (tokenOnJup) {
+    return {
+      symbol: tokenOnJup.symbol,
+      publicKey: tokenAddress.toString(),
+      url: tokenOnJup.logoURI,
+      decimals: tokenOnJup.decimals,
+    };
+  }
+
+  // next, try metaplex
+  const jsonMetadata = await getMetaplexMetadataForToken(
+    tokenAddress,
+    rpcProvider
+  );
+  if (jsonMetadata && jsonMetadata.symbol) {
+    return {
+      symbol: jsonMetadata.symbol,
+      publicKey: tokenAddress.toString(),
+      url: jsonMetadata.image,
+      decimals: jsonMetadata.seller_fee_basis_points,
+    };
+  }
+  let mint: Mint | undefined = undefined;
+  try {
+    //try getting metadata from token 2022
+    const token2022Metadata = await getTokenMetadata(
+      rpcProvider.connection,
+      tokenAddress
```

Comment:
So you have to pass the program_id for token 2022 otherwise this fetches tokenkeg, Which is fine we need / want both. Just that this by default is not token-2022, thats another program (and not all spl tokens are using token-2022, many are just the splprogram).

Reply:
Interestingly enough it has TOKEN_2022_PROGRAM_ID as the default in the library. But you know I think we should still add it anyway in case they change this in a later version and that could make this more brittle.
![image](https://github.com/metaDAOproject/futarchy-ts/assets/41474751/c20a3536-d8b0-43be-b32c-f744f651ea56)

Reply:
Okay added this